engines/sci/engine/kgraphics32.cpp

```diff
@@ -360,29 +360,112 @@ reg_t kScrollWindow(EngineState *s, int argc, reg_t *argv) {
+	r.top = readSelectorValue(segMan, object, SELECTOR(nsTop));
+	r.right = readSelectorValue(segMan, object, SELECTOR(nsRight)) + 1;
+	r.bottom = readSelectorValue(segMan, object, SELECTOR(nsBottom)) + 1;
+	Common::Point p(r.left, r.top);
+
+	ScrollWindow *scrollWindow = new ScrollWindow(segMan, r, p, plane,
+	                                              foreColor, backColor, fontId,
+	                                              (TextAlign)mode, borderColor);
+
+	return g_sci->_gfxControls32->registerScrollWindow(scrollWindow);
 }
 
 reg_t kScrollWindowAdd(EngineState *s, int argc, reg_t *argv) {
-	debug("kScrollWindowAdd");
-	return kStubNull(s, argc, argv);
+	ScrollWindow *scrollWindow = g_sci->_gfxControls32->getScrollWindow(argv[0]);
+	if (!scrollWindow)
+		error("Invalid ScrollWindow ID");
+
+	Common::String text = s->_segMan->getString(argv[1]);
+
+	bool scrollTo = true;
+	if (argc >= 6)
+		scrollTo = !argv[5].isNull();
+
+	return scrollWindow->add(text, argv[2].toSint16(), argv[3].toSint16(),
+	                  argv[4].toSint16(), scrollTo);
 }
 
 reg_t kScrollWindowWhere(EngineState *s, int argc, reg_t *argv) {
-	debug("kScrollWindowWhere");
-	return kStubNull(s, argc, argv);
+	ScrollWindow *scrollWindow = g_sci->_gfxControls32->getScrollWindow(argv[0]);
+	if (!scrollWindow)
+		error("Invalid ScrollWindow ID");
+
+	Common::Rational w = scrollWindow->where();
+
+	int ret = (argv[1].toUint16() * w).toInt();
+
+	return make_reg(0, ret);
 }
 
+reg_t kScrollWindowHide(EngineState *s, int argc, reg_t *argv) {
+	ScrollWindow *scrollWindow = g_sci->_gfxControls32->getScrollWindow(argv[0]);
+	if (!scrollWindow)
+		error("Invalid ScrollWindow ID");
+
+	scrollWindow->hide();
+
+	return s->r_acc;
+}
+
+
 reg_t kScrollWindowShow(EngineState *s, int argc, reg_t *argv) {
-	debug("kScrollWindowShow");
-	return kStubNull(s, argc, argv);
+	ScrollWindow *scrollWindow = g_sci->_gfxControls32->getScrollWindow(argv[0]);
+	if (!scrollWindow)
+		error("Invalid ScrollWindow ID");
+
+	scrollWindow->show();
+
+	return s->r_acc;
+}
+
+reg_t kScrollWindowUpArrow(EngineState *s, int argc, reg_t *argv) {
+	ScrollWindow *scrollWindow = g_sci->_gfxControls32->getScrollWindow(argv[0]);
+	if (!scrollWindow)
+		error("Invalid ScrollWindow ID");
+
+	scrollWindow->upArrow();
+
+	return s->r_acc;
 }
 
+reg_t kScrollWindowDownArrow(EngineState *s, int argc, reg_t *argv) {
+	ScrollWindow *scrollWindow = g_sci->_gfxControls32->getScrollWindow(argv[0]);
+	if (!scrollWindow)
+		error("Invalid ScrollWindow ID");
+
+	scrollWindow->downArrow();
+
+	return s->r_acc;
+}
+
+
 reg_t kScrollWindowDestroy(EngineState *s, int argc, reg_t *argv) {
-	debug("kScrollWindowDestroy");
-	return kStubNull(s, argc, argv);
+	ScrollWindow *scrollWindow = g_sci->_gfxControls32->getScrollWindow(argv[0]);
+	if (!scrollWindow)
+		error("Invalid ScrollWindow ID");
+
+	scrollWindow->hide();
+
+	g_sci->_gfxControls32->deregisterScrollWindow(argv[0]);
+	delete scrollWindow;
```

Comment:
As with the create call it seems like this could be fully encapsulated in GfxControls32, then memory control of the ScrollWindow object remains entirely within GfxControls32 instead of leaking out into kernel calls.